irgen: basic code generation for algebraic data types

eisenstatdavid opened the pull request:
Provides an API similar to the one specified in #16240.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16857)
<!-- Reviewable:end -->

knz commented on pkg/sql/ir/irgen/codegen/codegen.go:
```diff
@@ -0,0 +1,243 @@
+	if s.LenAlloc <= 0 {
+		panic("LenAlloc must be positive")
+	}
+	fmt.Fprintf(b, "package d\n")
```
please consider https://github.com/knz/compose-text

knz commented on pkg/sql/ir/irgen/test/main.go:
```diff
@@ -0,0 +1,71 @@
+}
+
+func main() {
+	var a Allocator
```
turn this into a Go test

knz reviewed (commented):
I really like the look & feel of the generated code!
seeing this result makes me more and more confident this is a good investment.

My concern right now is similar as during my first attempt at the end of last  year: the code for the printing is very verbose and poorly maintainable. I wonder if we can do better, I made a suggestion to this effect.

Also TC is complaining about a few things.

Commit:
irgen: analyzer requires exported names

Generated code will have its own package, so these names would not be
visible otherwise.

Commit:
irgen: introduce String methods for name types

Commit:
irgen: code generation for basic access and updates

Provides an API similar to the one specified in the RFC committed by
#16240. This commit updates that RFC to reflect the actual
implementation.

eisenstatdavid commented:
Code generation was completely rewritten. WIP because there is still significant documentation effort left: better comments, updating the RFC, and an explanation of the concepts behind `irgen/template`, but @knz this is ready for another look.

Commit:
rfc: update the algebraic data types rfc to reflect what was implemented

Some naming conventions were changed, and `Allocator` is passed by
value.

eisenstatdavid commented:
Added documentation and tidied the lint output. This is ready for formal review now, @knz.

knz commented:
This is very good. Thank you!
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 2 of 7 files at r2, 21 of 21 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16857#-:-KpVTMX_376tTzJq5Bfy:b-kx5jj4)*
<!-- Sent from Reviewable.io -->
